// eslint-plugin-deprecate/dist/rules/import.js
'use strict';

var ImportChecker = require('../ImportChecker');

module.exports = {
    meta: {
        docs: {
            description: 'Forbids importing from given files.'
        }
    },
    create: function create(context) {
        var imports = context.options.map(function (opt) {
            return new ImportChecker(opt);
        });
        return {
            ImportDeclaration: function ImportDeclaration(node) {
                var importPath = node.source.value;

                imports.filter(function (imp) {
                    return imp.check(importPath);
                }).forEach(function (imp) {
                    context.report({ node: node, message: imp.getErrMessage() });
                });
            },
            CallExpression: function CallExpression(node) {
                if (node.callee.name !== 'require' || !node.arguments.length) {
                    return;
                }
                var requireArg = node.arguments[0];
                if (requireArg.type !== 'Literal') {
                    return;
                }
                imports.filter(function (imp) {
                    return imp.check(requireArg.value);
                }).forEach(function (imp) {
                    context.report({ node: node, message: imp.getErrMessage() });
                });
            }
        };
    }
};
